source/common/json/json_loader.cc

```diff
@@ -2,242 +2,669 @@
+        break;
+      }
+      case Type::Boolean:
+        value.PushBack(element->value_.boolean_value_, allocator);
+        break;
+      case Type::Double:
+        value.PushBack(element->value_.double_value_, allocator);
+        break;
+      case Type::Integer:
+        value.PushBack(element->value_.integer_value_, allocator);
+        break;
+      case Type::Null:
+        value.PushBack(rapidjson::Value(), allocator);
+        break;
+      case Type::String:
+        value.PushBack(rapidjson::StringRef(element->value_.string_value_.c_str()), allocator);
+      }
     }
-    return object_array;
+    break;
   }
-
-  std::string getString(const std::string& name) const override {
-    rapidjson::Value::ConstMemberIterator member_itr = value_.FindMember(name.c_str());
-    if (member_itr == value_.MemberEnd() || !member_itr->value.IsString()) {
-      throw Exception(fmt::format("key '{}' missing or not a string in '{}'", name, name_));
+  case Type::Object: {
+    value.SetObject();
+    for (const auto& item : field.value_.object_value_) {
+      auto name = rapidjson::StringRef(item.first.c_str());
+
+      switch (item.second->type_) {
+      case Type::Array:
+      case Type::Object: {
+        rapidjson::Value nested_value;
+        buildRapidJsonDocument(*item.second, nested_value, allocator);
+        value.AddMember(name, nested_value, allocator);
+        break;
+      }
+      case Type::Boolean:
+        value.AddMember(name, item.second->value_.boolean_value_, allocator);
+        break;
+      case Type::Double:
+        value.AddMember(name, item.second->value_.double_value_, allocator);
+        break;
+      case Type::Integer:
+        value.AddMember(name, item.second->value_.integer_value_, allocator);
+        break;
+      case Type::Null:
+        value.AddMember(name, rapidjson::Value(), allocator);
+        break;
+      case Type::String:
+        value.AddMember(name, rapidjson::StringRef(item.second->value_.string_value_.c_str()),
+                        allocator);
+        break;
+      }
     }
-    return member_itr->value.GetString();
+    break;
   }
+  default:
+    throw Exception("Json has a non-object or non-array at the root.");
+  }
+}
 
-  std::string getString(const std::string& name, const std::string& default_value) const override {
-    if (!value_.HasMember(name.c_str())) {
-      return default_value;
-    } else {
-      return getString(name);
-    }
+rapidjson::Document Field::asRapidJsonDocument() const {
+  rapidjson::Document document;
+  rapidjson::Document::AllocatorType& allocator = document.GetAllocator();
+  buildRapidJsonDocument(*this, document, allocator);
+  return document;
+}
+
+uint64_t Field::hash() const {
+  rapidjson::StringBuffer buffer;
+  rapidjson::Writer<rapidjson::StringBuffer> writer(buffer);
+  asRapidJsonDocument().Accept(writer);
+  return std::hash<std::string>{}(buffer.GetString());
+}
+
+bool Field::getBoolean(const std::string& name) const {
```

Comment:
Can we add TODO here to add line number info to all these errors (or just do now)